dali/operators/random/normal_distribution_op.cu

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/random/normal_distribution_op.cuh"
+#include <vector>
+#include <utility>
+#include "dali/core/convert.h"
+
+namespace dali {
+
+namespace detail {
+
+template <typename Out, typename RandType>
+__global__ void NormalDistKernel(NormalDistributionGpu::BlockDesc *block_descs,
```

Comment:
I wonder how this kernel would compare with a simpler:
```

template <typename T>
__global__ void RandKernel(const SampleDesc* sample_descs, RandomizerGPU randomizer) {
  const int sample_idx = blockIdx.y;
  auto& sample = sample_descs[sample_idx];
  auto out = static_cast<Out*>(sample.out);
  
  const int64_t block_size = blockDim.y * blockDim.x;
  const int64_t grid_size = gridDim.x * block_size;
  const int64_t offset = block_size * blockIdx.x;
  const int64_t tid = threadIdx.y * blockDim.x + threadIdx.x;
  auto norm = randomizer.normal<RandType>(tid);
  for (int64_t idx = offset + tid; idx < sample.size; idx += grid_size) {
    out[idx] = ConvertSat<Out>(norm * sample.std + sample.mean);
  }
}
```
The benefit is that you'll get rid of the block calculation/distribution logic and the cuda memcpy for the block descriptors (you'll have only sample descriptors which is smaller). Recently we saw that those cudaMemcpyAsync calls for descriptors might impact the overall performance of a pipeline if they interfere with other memcpys so it is worth considering I think. You can easily write some benchmarks (see for instance crop_bench.cc for reference)

Reply:
I did some benchmarks and the performance of his kernel is worse, than what is there now. For big batches of ununiformly sized tensors it's much slower (up to 50%), for big uniformly sized batches the performance is slightly better in the current implementation (probably less coords calculation inside kernel). 
Also the drawback of copying the block descriptors might not be very big because there is a limited number of blocks, so the copy is smaller than in some DALI kernels. 
I'll stay with what I have, thanks for the suggestion, though.

Reply:
Thanks for checking that!